Update JS library (jquery & mathjax) using GitHub action

krmanik opened the pull request:
## Pull Request template

## Purpose / Description
The jquery and mathjax library should be updated and similar to Anki.

## Fixes
Fixes #11211

## Approach
1. Used GitHub actions to run bash script
2. Create a new branch for updating the js library then create PR from the branch
3. Use [anki](http://github.com/ankitects/anki) repository to build only for vendor
```sh
cd ~/tmp/anki/qt/aqt/data/web/js/vendor; bazel build vendor
```
4. Then copy jquery to assets dir
```sh
cp ~/tmp/anki/.bazel/bin/qt/aqt/data/web/js/vendor/jquery.min.js AnkiDroid/src/main/assets/jquery.min.js
```
5. Also, remove mathjax dir from assets dir and copy new mathjax dir to assests dir. (It is same as AnkiDesktop)
```sh
# remove old mathjax file
rm -rf AnkiDroid/src/main/assets/mathjax
mkdir AnkiDroid/src/main/assets/mathjax

# copy latest mathjax to assets dir
cp -r ~/tmp/anki/.bazel/bin/qt/aqt/data/web/js/vendor/mathjax AnkiDroid/src/main/assets/
```

## How Has This Been Tested?
Tested on local repository 
**Actions result**
https://github.com/krmanik/Anki-Android-Copy/runs/6649982980?check_suite_focus=true

**PR result**
https://github.com/krmanik/Anki-Android-Copy/pull/3
<img src="https://user-images.githubusercontent.com/12841290/170940925-f9ea7a6c-9626-4295-a252-83b84f85d307.png" width=450></img>

### Tested on device with new files
![image](https://user-images.githubusercontent.com/12841290/171459576-880e3fb7-f401-43d6-a59b-7de5bd0423a6.png)


## Learning (optional, can help others)
https://octokit.github.io/rest.js/v18#pulls-create

_Links to blog posts, patterns, libraries or addons used to solve this problem_

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

mikehardy commented on .github/workflows/update_js_libs.yml:
```diff
@@ -0,0 +1,56 @@
+        id: vars
+        run: |
+
+          # check if branch exists on remote
+          branch_in_remote=$(git ls-remote --heads origin update-js-libs)
+          if [[ -n ${branch_in_remote} ]]; then
+            git push origin --delete update-js-libs
+          fi
+
+          # create a new branch and checkout on local
+          git branch update-js-libs
+          git checkout update-js-libs
```
Hmm :thinking: - in the i18n_sync branch I don't care about it's existing state and do not attempt to delete it, I just force-push it at the end. Is that an option here? Might simplify things by not having to care about state, and I think a force-push is a fraction safer than a branch delete / new branch - with github we will still have a record of the old commit hashes on the branch if need be. I think simplifying the code is a stronger argument then safety though

mikehardy commented on .github/workflows/update_js_libs.yml:
```diff
@@ -0,0 +1,56 @@
+              head: 'update-js-libs',
+              base: 'main',
+              title: 'Update jquery and mathjax',
+              body: "### Updated\n ${{ steps.vars.outputs.jquery_ver }} \n ${{ steps.vars.outputs.mathjax_ver }}"
```
that's a nice touch, with the versions

mikehardy reviewed (changes requested):
This is really cool! You are turning into Sir Automation, Kinght of the Github Action Round Table :-) - seriously, this stuff helps, thank you

1- One question on branch delete/re-create
2- this looks like a build artifact that could be removed? https://github.com/krmanik/Anki-Android-Copy/pull/3/files#diff-8a2d33e819daa825c5453bde83d1d41e16d095c75456a9bb1b9b472972ac6957
3- it seems like most of the diff is file mode changes, I guess that is just how bazel builds things? with file mode 755 instead of 644? These files really shouldn't be executable though - I wonder if it is possible to make them 644 again, then the resulting PR would be much smaller. At the same time, it's not that important and I don't want to modify the upstream build process too much, if really wants them to be 755 it will only be a diff for file mode one time

krmanik commented:
I have implemented this python script for doing the same task. But I think bazel build should be used because in next version of Anki if another library added then the build will/may provide the js files.

<details>
  <summary>Click to view update-js-libs.py</summary>

```js
#!/usr/bin/python3

import os
import shutil
import tarfile
import requests


assets_directory = "AnkiDroid/src/main/assets/"

""" 
Update jquery
1. Check latest version using npm registry
2. Get latest jquery.min.js file from official website
3. Save js file to assets directory
"""
def update_jquery():
    # get version details from npm registry
    url = "https://registry.npmjs.org/jquery/latest"
    req = requests.get(url)
    data = req.json()
    version = data["version"]
    
    # get latest jquery for official website
    js_url = "https://code.jquery.com/jquery-" + version + ".min.js"
    req = requests.get(js_url)
    js = req.text

    # save latest jquery lib to assets directory
    jquery_js_file = open(assets_directory + "jquery.min.js", "w")
    jquery_js_file.write(js)


"""
Update mathjax
1. Download latest tar.gz file using npm registry
2. Extract and copy selected file to assets/mathjax directory
"""
def update_mathjax():
    # get version details from npm registry
    url = "https://registry.npmjs.org/mathjax/latest"
    req = requests.get(url)
    data = req.json()
    tgz = data["dist"]["tarball"]

    # get tar.gz file
    tgz_file = open("mathjax.tgz", "wb")
    req = requests.get(data["dist"]["tarball"], stream=True)
    req.raw.decode_content = True
    shutil.copyfileobj(req.raw, tgz_file)
    tgz_file.close()

    # extract tgz file to current directory
    tar = tarfile.open("mathjax.tgz", "r:gz")
    tar.extractall()
    tar.close()

    # copy selected files in mathjax dir to assets dir
    with open("tools/mathjax-cp.txt", "r") as f:
        lines = f.readlines()
        mathjax_dir = assets_directory + "mathjax"
        for line in lines:
            src = line.strip()
            dst = mathjax_dir + line[11:].strip()
            shutil.copyfile(src, dst)
    
    # remove temp files
    os.remove("mathjax.tgz")
    shutil.rmtree('package', ignore_errors=True)

update_jquery()
update_mathjax()
```  

</details>

For mode of file I have tried following but it is not working as expected. On local system it works.
1. `git update-index --add --chmod=-x file-name`
2. `chmod -x file-name`
3. `chown` then `chmod`
May be I need help here or I should try again with other implementation.

The branch create/delete will be changed and re-implemented similar to  `i18n_sync` branch.

mikehardy commented:
I agree sticking with the upstream build style is best, that way even if it changes we can just refer to upstream build documentation to adapt to the changes

If the file modes are hard to change, let's just go with it. That said, after the build a `chmod -R 644 <files>` should do it? Of course if it works on a local system but not in CI then maybe there is something subtle going on. It may not be worth but one more test then just accepting it and moving on

krmanik commented:
Then I will continue with `bazel build` and one more try for mode of file then push the changes.

Also one things to note, this commit in MathJax may be need to address (in another PR) because from MathJax 3.2.1 the Locale is change from JS to JSON. So loading JSON may give errors in webview (untested).

From release notes of [MathJax 3.2.1](https://github.com/mathjax/MathJax-src/releases/tag/3.2.1)
`Locale files are now served with a .json extension. (https://github.com/mathjax/MathJax/issues/2403)`

mikehardy commented:
I guess we will see the Locale JSON affect when we get the PR from this new automation :-)

krmanik commented:
Updated with following
1. Restored file mode in assets dir
2. Changed `_anki/js/vendor/mathjax` to `/android_asset\/mathjax/ in `conf.js`
3. If PR exists then force push to the PR otherwise new PR created

Also build apk with new files and tested on devices, no errors in dev tools console the mathjax files loaded successfully.

Arthur-Milchior reviewed (commented):
Really cool.

I am not competent enough in github action to review it. But since it's an action for dev only, and if it breaks it creates no issue to anyone, I'm cool with just trusting the fact that you tested it.

I'd like to note that it'll break the day when dae moves its folder structure. Given history, I'm quite certain it'll occur again. As I stated, no big deal, at worst the script break.
I suspect it may be easier to update if you put the Anki folder into variables. Potentially even the ankidroid folders just in case we change it one day. So that the part we expect to maybe change is clear and separated from the execution part

Thanks again

krmanik commented on tools/update-js-libs.sh:
```diff
@@ -0,0 +1,49 @@
+ANKI_MATHJAX_DIR="_anki/js/vendor/mathjax"
+ANKIDROID_MATHJAX_DIR="android_asset/mathjax"
+
+sed -i "s|$ANKI_MATHJAX_DIR|$ANKIDROID_MATHJAX_DIR|" $ANKIDROID_SRC/AnkiDroid/src/main/assets/mathjax/conf.js
```
First check if string present in dir preset in the file or not.

krmanik commented:
For 5 minutes manual task I spent more than 5 hours to implement this automation :)

Arthur-Milchior commented on tools/update-js-libs.sh:
```diff
@@ -8,37 +12,38 @@ curl -L https://github.com/bazelbuild/bazelisk/releases/download/v1.10.1/bazelis
 
 # build for mathjax and jquery
-cd ~/tmp/anki/qt/aqt/data/web/js/vendor
+cd $ANKI_SRC/qt/aqt/data/web/js/vendor
```
Oops, my bad for unclarity. 
It was path such as `qt/aqt/data/web/js/vendor` that I thought we'd put in variable, for when DAE change it.

Anyway, I'm fine with it, and delegating to the day where this is actually changed

mikehardy reviewed (approved):
This looks great - the PR on your fork looks good as does the action run :+1:
